### ISSUE TITLE
Image inpainting using Stable Diffusion

### ISSUE THREAD

**Issue by dbuscombe-usgs** [Doodleverse/CoastSeg#144]
Problem: gaps in imagery caused by clouds and other artifacts cause discontinuous shorelines and loss of data

Solution: I have implemented a Stable Diffusion based image inpainter based on runwayml/stable-diffusion that inpaints an image automatically. It mimics the behaviour of https://huggingface.co/spaces/runwayml/stable-diffusion-inpainting without the manual mask making. Instead, it generates a mask by dilating the zero-mask of the image's red channel. The image ans mask are passed to the `StableDiffusionInpaintPipeline` with custom hyperparameters determined through research on a set of satellite imagery from a variety of locations.

In an activated coastseg environment:

```
pip install accelerate diffusers
```
(I'm not 100% sure if `accelerate` is required)

these are the required imports 
```
from diffusers import StableDiffusionInpaintPipeline
import torch
from PIL import Image
from skimage.morphology import binary_dilation, disk
import numpy as np
from glob import glob
```

Set up the model pipeline for GPU

```
pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting",
    revision="fp16",
    torch_dtype=torch.float16,
).to("cuda")
```

for CPU, I think it has to be

```
pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting"
)
```

The model seems to want 512x512 imagery, so I resize the imagery to 512x512 for the model, then back to the original size before writing to file. 
```
N=512
```

I found that the model works best if the prompt is blank. The threshold is just for masking out zeros
```
thres = 0.01
prompt =""
```

Loop through some images, writing out to file with suffix "inpaint.jpg"
```
files = glob("*.jpg")

for file in files:
    # open file and get size
    image_orig = Image.open(file)
    nx, ny = image_orig.size
    ## resize to NxN
    image_orig = image_orig.convert("RGB").resize((N,N))
    ## create mask from from channel zeros
    mask = ((np.array(image_orig)[:,:,0]<thres) * 255).astype(np.uint8)
    ## dilate the mask, to give the inpainter some pixels to work with
    mask = binary_dilation(mask, disk(9))
    ## convert back to PIL format
    mask_image = Image.fromarray(mask)
    ##I found a large guidance scale is required. I'm still playing with optimal parameters
    output = pipe(prompt = prompt, image=image_orig, mask_image=mask_image, num_inference_steps=100, guidance_scale=9)
    ## resize output and write to disk
    output = output.images[0].resize((nx,ny))
    output.save(file.replace(".jpg","_inpaint.jpg"))
```

Full implementation example script here 
[inpaint_satRGBimage.zip](https://github.com/Doodleverse/CoastSeg/files/11644566/inpaint_satRGBimage.zip)

Examples:

![2017-12-29-15-42-51_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/79dffefb-6740-482c-9f06-1713d928c15f)
![2017-12-29-15-42-51_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/fe63cc3d-279f-4b7b-a2a8-d77c7811b7a7)

![Elwha_2019-02-17-19-01-10_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/3596509/66e6fc1b-f465-4f0d-a814-e8d855066c13)
![Elwha_2019-02-17-19-01-10_RGB_L8_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/3b502f69-7c8a-42fb-9eef-0b1d08aff789)

![2018-01-06-15-40-23_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/3596509/a1c54241-6466-43fa-9408-b630ba01ba9b)
![2018-01-06-15-40-23_RGB_L8_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/d4f4a08c-454d-4537-9533-dcff10557317)

![2009-07-15-21-14-06_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/99224321-7d95-40f8-b98d-f6472c706804)
![2009-07-15-21-14-06_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/12234d32-88c4-4311-a84c-b0b7b5ec67de)

![Elwha_2019-01-07-19-07-28_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/3596509/8b529c7b-73b1-4542-b401-c90116794947)
![Elwha_2019-01-07-19-07-28_RGB_L8_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/76c3da3a-0cb2-4efe-ae2a-59d06f854b1f)

![2003-02-04-20-42-40_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/6d3fc4b1-64d3-405b-9f3a-850e2b8caab0)
![2003-02-04-20-42-40_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/f4485812-bd0b-4cea-94d6-a284435b9473)

**Comment by dbuscombe-usgs** [comment]
![2014-08-23-21-15-13_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/d4b2c9a3-7b53-43f5-8624-e7d2824b5610)
![2014-08-23-21-15-13_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/e888d9ae-c5b9-4854-8dd0-08cafd4471f6)

![2011-03-17-15-23-33_RGB_L5](https://github.com/Doodleverse/CoastSeg/assets/3596509/3db8fc60-adb9-4b2e-b5cb-0e5e8dda9c6c)
![2011-03-17-15-23-33_RGB_L5_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/8311c5ed-1aa5-405c-b3c6-87a6289da7f2)

![2011-08-22-21-17-10_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/06be64cb-d764-493c-bc3d-3d8c06025b23)
![2011-08-22-21-17-10_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/c82fe349-bf89-4c26-88ab-c23c37d44587)

![2018-06-25-16-00-37_RGB_S2](https://github.com/Doodleverse/CoastSeg/assets/3596509/6f986a4d-066a-4425-b70c-5a54d828702f)
![2018-06-25-16-00-37_RGB_S2_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/50ff7bd9-28da-4a4a-ac24-f8a4481ebe9c)

**Comment by dbuscombe-usgs** [comment]
It does struggle on some occasions -  I guess the gaps can be too large! A new glitch art form

perhaps this combined with a % black filter would work well, and it does work well most of the time!

![2002-07-05-15-16-23_RGB_L5](https://github.com/Doodleverse/CoastSeg/assets/3596509/e09bb59d-6940-483e-b985-af755db9c26a)
![2002-07-05-15-16-23_RGB_L5_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/62aa3dc1-c9df-4813-b918-73836f68fc17)

![2004-01-08-15-29-11_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/211b5559-6b62-4ae5-9e76-d256aa40c206)
![2004-01-08-15-29-11_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/b89e7ee8-a48f-49e9-a525-17b2bd022ec1)

![2008-07-21-15-26-55_RGB_L5](https://github.com/Doodleverse/CoastSeg/assets/3596509/6e5ae0eb-e9e9-4260-a78f-f16e873c3d1d)
![2008-07-21-15-26-55_RGB_L5_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/18ea8d5c-ea7a-4cf3-9230-16dbffc65892)


![2010-09-21-15-32-33_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/c3e7fa50-811c-4a5c-ab23-044c6c83fb0e)
![2010-09-21-15-32-33_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/50bf836c-7930-485c-94bb-0d14c0ec75df)

![2011-05-03-15-34-16_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/e611d841-c402-4375-8b9f-f140ef153f9f)
![2011-05-03-15-34-16_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/45fe446f-79c7-42ab-b4ac-a523ff049149)

![2011-11-11-15-33-48_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/079d7a1e-9b43-43c4-a91b-5d809897124d)
![2011-11-11-15-33-48_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/0bb6a0b8-b70e-4b6d-aba6-88cb35820eba)

![2014-06-20-15-39-59_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/3596509/d8d4155f-2402-44a1-965e-92b5bcfe5963)
![2014-06-20-15-39-59_RGB_L8_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/db360fa6-65eb-44c0-a581-196cdefb83d8)


![2018-06-16-15-34-10_RGB_L7](https://github.com/Doodleverse/CoastSeg/assets/3596509/e167cbea-4610-4e1c-9207-41d9edfa0a04)
![2018-06-16-15-34-10_RGB_L7_inpaint](https://github.com/Doodleverse/CoastSeg/assets/3596509/6b34244a-4450-48c6-92ed-22025f09d5c8)

**Comment by dbuscombe-usgs** [comment]
I modified the code so it only replaces previously black pixels, and fills the gaps based on a median of 3 realizations of the inpainted image using stable diffusion. Latest version here [inpaint_satRGBimage.zip](https://github.com/Doodleverse/CoastSeg/files/11645977/inpaint_satRGBimage.zip)

check out the results ! 
Noisy: https://drive.google.com/file/d/1vRQlKh55jV5Jr8Rj_YgibecOMvlQ6nNh/view?usp=sharing
Clean: https://drive.google.com/file/d/1fr4NXKv07WSVpMmaAHj5ninpilhbrME2/view?usp=sharing

**Comment by dbuscombe-usgs** [comment]
Let's just add this later https://github.com/dbuscombe-usgs/SatelliteInPaint

closing for now
